Addresses FCREPO-1875

emetsger opened the pull request.

ajs6f commented:
Nice, nice. One question: does `HttpResponse` itself not come in a `Closeable` form? Maybe it would be better to use that underneath `FcrepoResponse` and let the HTTP machinery handle closing resources inside itself?

acoburn commented:
Very nice. Am I correct that this supersedes #6?

emetsger commented:
@ajs6f Ah, yes it does, pushing an update.  @acoburn yes this supersedes #6.

There may also be a bug lurking in `getEntityContent(InputStream)` but I'm not sure how easy that is to test.  If an IOE is thrown from `HttpEntity.getContent()` it's unclear to me what the `HttpClient` will do with the resources associated with the `HttpResponse`.  I'm not really sure if that's worth the effort to investigate as a part of this particular issue.

ajs6f commented:
Not sure that it is, but I reckon it would result in the addition of a `finally` clause, anyway.

awoods commented:
@emetsger, could you please create a JIRA ticket for this?

emetsger commented:
Hi Andrew-

Sorry I don't understand what you mean, exactly.

https://jira.duraspace.org/browse/FCREPO-1875 exists tracking this issue.

Do you just want a Jira issue created for the PR specifically?

On Wednesday, February 3, 2016, Andrew Woods <notifications@github.com>
wrote:

> @emetsger <https://github.com/emetsger>, could you please create a JIRA
> ticket for this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4-exts/fcrepo-java-client/pull/7#issuecomment-179017365>
> .
>

acoburn commented:
@emetsger you're all set: fcrepo-1875 takes care of this.

awoods commented:
@emetsger, thanks for the link. For bookkeeping reasons, I find it most helpful to track in-flight github work with an associated JIRA ticket... hence the request for a link in the comments of this PR. Likewise, it is helpful to have a link to the PR over in the JIRA ticket itself, along with the current "Assignee" and status of the work ("In Progress", "In Review", etc).
Here is a description of the process:
https://wiki.duraspace.org/display/FF/Story+Management

emetsger commented:
@awoods ok!  I just want to be sure I'm following process:
1. This PR (#7) now has a link (in the [comments](https://github.com/fcrepo4-exts/fcrepo-java-client/pull/7#issuecomment-179866657)) to the Jira issue FCREPO-1875.
2. The [Jira issue](https://jira.duraspace.org/browse/FCREPO-1875) has a [direct link](https://jira.duraspace.org/browse/FCREPO-1875?focusedCommentId=47735&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-47735) to [this PR](#7)

awoods commented:
Thanks, @emetsger. Sorry for the run-around.

ajs6f commented on src/main/java/org/fcrepo/client/FcrepoResponse.java:
```diff
@@ -55,6 +72,36 @@ public FcrepoResponse(final URI url, final int statusCode,
+     *
+     * @return {@code true} if resources have been freed, otherwise {@code false}
+     */
+    public boolean isClosed() {
```
Is there any specific reason to expose this?

emetsger replied:
My thought was allowing a client to avoid accessing the underlying closed response (or exhausted `InputStream`), which may result in an exception.  Not wedded to it, though, and whether or not this needs to be exposed may be influenced by factoring up the `CloseableHttpResponse`.

ajs6f replied:
Hm. Not sure what to think. I guess it doesn't hurt much...

ajs6f commented on src/main/java/org/fcrepo/client/FcrepoResponse.java:
```diff
@@ -55,6 +72,36 @@ public FcrepoResponse(final URI url, final int statusCode,
+     * @throws IOException if there is an error closing the underlying HTTP response stream.
+     */
+    @Override
+    public void close() throws IOException {
```
Do we want to keep a ref directly to the underlying response so that we can `close` it? It's possible (even if it is not the case now) that so doing would release other resources inside it (besides what we here call `body`).

emetsger replied:
Sure, that makes sense.  Should that be done in this PR?  Happy to do it.

ajs6f replied:
Yeah, why not. And I would get rid of `body` and just call into that underlying response directly.

emetsger replied:
While I agree that this work should be done, I'm convinced this should be handled in another Jira issue.  Primarily because of the amount of work involved: `setBody(InputStream)` on the `FcrepoReponse` is used a lot in the tests.  If the `InputStream body` member is going away in favor of a `CloseableHttpResponse` member, then `setBody(InputStream)` no longer makes sense to keep.  This is also tied into making `FcrepoResponse` immutable, which probably shouldn't be a part of this work either.

Secondarily, the aim of this PR is to really make it clear who is responsible for resource management, and that's been accomplished.  Using `CloseableHttpResponse` in `FcrepoResponse` is a (worthy) optimization, as is an immutable `FcrepoResponse`, but I'm not sure they belong in this PR.

awoods replied:
@emetsger, could you please create this follow-on JIRA ticket?

ajs6f replied:
Okay, but can you open another ticket for these things, so that we don't lose track of them?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+         * number of connections.
+         */
+        static Consumer<HttpClientConnectionManager> connectionRequestedAndClosed(final int connectionCount) {
+            return (connectionManager) -> {
```
No need for parens without typing the param.

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+    /**
+     * Verifications for HttpClientConnectionManager
+     */
+    static class Verifications {
```
Can this not be `private`?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+         * number of connections, but hasn't closed <em>any</em> of them
+         */
+
+        static Consumer<HttpClientConnectionManager> connectionRequestedButNotClosed(final int connectionCount) {
```
Can this not be `private`?

ajs6f replied:
It's not totally clear to me where the value of making this a `Consumer` (and not just a `static void` method with a `HttpClientConnectionManager` param) is coming in?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+    /**
+     * FcrepoResponse handlers.
+     */
+    static class FcrepoResponseHandler {
```
Can this not be `private`?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+        /**
+         * Closes the InputStream that constitutes the response body.
+         */
+        static Consumer<FcrepoResponse> closeEntityBody = (response) ->
```
Can this not be `private`?

ajs6f replied:
No need for parens without typing the param.

ajs6f replied:
It's not totally clear to me where the value of making this a `Consumer` (and not just a `static void` method with a `FcrepoResponse` param) is coming in?

emetsger replied:
The `Consumer` is a uniform interface and semantics for the finally block in `connect(...)`.  Tests can pass in the appropriate Consumer for their needs.

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+        /**
+         * Reads the InputStream that constitutes the response body.
+         */
+        static Consumer<FcrepoResponse> readEntityBody = (response) -> {
```
Can this not be `private`?

ajs6f replied:
No need for parens without typing the param.

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+         */
+
+        static Consumer<HttpClientConnectionManager> connectionRequestedButNotClosed(final int connectionCount) {
+            return (connectionManager) -> {
```
No need for parens without typing the param.

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+    private PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
+
+    @Before
+    public void setUp() throws Exception {
```
Can this `throw` something more specific?

ajs6f commented on src/main/java/org/fcrepo/client/FcrepoClient.java:
```diff
@@ -268,11 +292,27 @@ private FcrepoResponse fcrepoGenericResponse(final URI url, final HttpResponse r
+     * @param response the response object to close
+     * @throws IOException if there is trouble closing the {@code response}
+     */
+    private void free(final CloseableHttpResponse response) throws IOException {
```
You never do anything with this `IOException`. Maybe swallow it here?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    @After
+    public void tearDown() throws Exception {
```
Can this `throw` something more specific?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+     * @throws Exception if something exceptional happens
+     */
+    @Test
+    public void connectionReleasedOnException() throws Exception {
```
This test and the test following are written very differently. Even if the atomic machinery isn't needed here, it might be best to write this test with it to keep some uniformity.

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+        assertEquals("Expected to make " + expectedCount + " connections; made " + actualCount,
+                expectedCount, actualCount);
+
+        Verifications.connectionRequestedAndClosed(actualCount).accept(connectionManager);
```
Maybe use a `static import` here.

emetsger replied:
The reason I didn't is because `Verifications.` is essentially a comment disguised as a class name, but happy to change.

ajs6f replied:
Eh, it's not huge. But I think it might be clearer to scrap `Verifications` and just have a bunch of methods with a shared prefix in the name.

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+         * @return a Consumer that verifies the supplied HttpClientConnectionManager has opened and closed the expected
+         * number of connections.
+         */
+        static Consumer<HttpClientConnectionManager> connectionRequestedAndClosed(final int connectionCount) {
```
It's not totally clear to me where the value of making this a `Consumer` (and not just a `static void` method with a `HttpClientConnectionManager` param) is coming in?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+        final int expectedCount = (int) Stream.of(HttpMethods.values()).filter(m -> m.entity).count();
+        final AtomicInteger actualCount = new AtomicInteger(0);
+        final MockHttpExpectations.Uris uri = uris.uri200;
+        final Consumer<FcrepoResponse> responseHandler = FcrepoResponseHandler.closeEntityBody;
```
Why take a ref to a `static` member?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+        final int expectedCount = (int) Stream.of(HttpMethods.values()).filter(m -> m.entity).count();
+        final AtomicInteger actualCount = new AtomicInteger(0);
+        final MockHttpExpectations.Uris uri = uris.uri200;
+        final Consumer<FcrepoResponse> responseHandler = FcrepoResponseHandler.readEntityBody;
```
Why take a ref to a `static` member?

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+
+                case OPTIONS:
+                    // not currently supported by the FcrepoClient
+                    for (Method m : client.getClass().getDeclaredMethods()) {
```
Please note in the comments that this loop is an intentional time bomb.

ajs6f commented:
There are a lot of tests here that `throw Exception`. It would be nice if they would throw something more specific.

ajs6f commented on src/test/java/org/fcrepo/client/FcrepoResponseTest.java:
```diff
@@ -69,4 +79,103 @@ public void testResponse() throws IOException {
+     * exception is thrown inside of a client's {@code try} block, the {@code IOException} from the {@code close()}
+     * method is properly appended as a suppressed exception.
+     *
+     * @throws Exception if something exceptional happens
```
This is pretty much why I prefer tests to `throw` something more specific than `Exception`.

ajs6f commented on src/test/java/org/fcrepo/client/FcrepoResponseTest.java:
```diff
@@ -69,4 +79,103 @@ public void testResponse() throws IOException {
+            throw new RuntimeException(notSuppressedMsg);
+
+        } catch (Exception e) {
+            assertEquals(notSuppressedMsg, e.getMessage());
```
Instead of comparing messages between two constructed objects, you might throw a specific object from the mock and use a direct `==` comparison. I think that might be a bit clearer.

ajs6f commented on src/test/java/org/fcrepo/client/FcrepoResponseTest.java:
```diff
@@ -69,4 +79,103 @@ public void testResponse() throws IOException {
+
+        try (FcrepoResponse res = client.get(URI.create("foo"), "", "")) {
+            assertEquals(content, IOUtils.toString(res.getBody()));
+        } catch (Exception e) {
```
This is... quite general? Is there a specific `Exception` that is actually likely here?

ajs6f commented on src/test/java/org/fcrepo/client/FcrepoResponseTest.java:
```diff
@@ -69,4 +79,103 @@ public void testResponse() throws IOException {
+        try (FcrepoResponse res = client.get(URI.create("foo"), "", "")) {
+            IOUtils.copy(res.getBody(), NullOutputStream.NULL_OUTPUT_STREAM);
+            fail("Expected an IOException to be thrown.");
+        } catch (Exception e) {
```
Maybe `catch` just `IOException` here. We want anything else to escape.

ajs6f commented on src/test/java/org/fcrepo/client/MockHttpExpectations.java:
```diff
@@ -0,0 +1,117 @@
+    static String port;
+
+    static class Uris {
+        int statusCode;
```
`final`?

ajs6f commented on src/test/java/org/fcrepo/client/MockHttpExpectations.java:
```diff
@@ -0,0 +1,117 @@
+
+    static class Uris {
+        int statusCode;
+        String path;
```
`final`?

ajs6f commented on src/test/java/org/fcrepo/client/MockHttpExpectations.java:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    static class SupportedUris {
```
This type looks rather like an `enum`. Maybe make it one?

emetsger replied:
The issue is there is state in the Uris class, especially port number, that is only assigned after the `MockHttpExpectations().initializeExpectations(...)` is called.

ajs6f commented:
Our normal practice is definitely to squash before merging, but I can do that for you if you want.

ajs6f commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,371 @@
+        static Consumer<FcrepoResponse> readEntityBody = (response) -> {
+            assertNotNull("Expected a non-null InputStream.", response.getBody());
+            try {
+                IOUtils.copy(response.getBody(), NullOutputStream.NULL_OUTPUT_STREAM);
```
Prefer Guava `ByteStreams.copy`.

ajs6f commented on src/test/java/org/fcrepo/client/FcrepoResponseTest.java:
```diff
@@ -69,4 +79,103 @@ public void testResponse() throws IOException {
+                new FcrepoResponse(null, 200, null, null, mockBody));
+
+        try (FcrepoResponse res = client.get(URI.create("foo"), "", "")) {
+            IOUtils.copy(res.getBody(), NullOutputStream.NULL_OUTPUT_STREAM);
```
Prefer Guava `ByteStreams::copy`.

ajs6f commented:
I left a lot of comments on this, @emetsger, and a lot of them are pretty straightforward. Let me know if you want a PR to your PR to address some of them. I know you have plenty to do and we appreciate you bringing in this PR at all.

emetsger commented:
@ajs6f no worries!  Very thorough review ;)  I'll update later today-

emetsger commented:
@ajs6f I addressed most of your comments with commits 3aed34d..164288c.  The ones I didn't address are those that I've pushed back on (and obviously can be discussed further):

* Replacing the member `InputStream body` with `ClosableHttpResponse` in `FcrepoResponse`.
* Use of a `Consumer` for the `FcrepoResponseHandler`.
* The `close()` method on `FcrepoResponse`.
* Refactoring the `SupportedUris`/`Uris` class to an `enum` within the `MockHttpExceptions` class.

awoods commented on src/main/java/org/fcrepo/client/FcrepoClient.java:
```diff
@@ -268,11 +284,26 @@ private FcrepoResponse fcrepoGenericResponse(final URI url, final HttpResponse r
+        try {
+            response.close();
+        } catch (IOException e) {
+            // No viable recovery, so we'll swallow it.
```
A simple warning log message may be informative, if not helpful.

emetsger replied:
WARN level or INFO level?

ajs6f replied:
I'd say at least `WARN`. An `IOException` here is _not_ normal operation.

awoods replied:
Sure, let's do INFO.

awoods replied:
Given @ajs6f's response... shall we stick with WARN?

emetsger replied:
WARN it is!

awoods commented on src/test/java/org/fcrepo/client/ConnectionManagementIT.java:
```diff
@@ -0,0 +1,354 @@
+ * @author esm
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class ConnectionManagementIT {
```
Typically, we have been calling tests "integration-tests" when they interact with a running repository, and "unit-tests" when the interactions are mocked. I would suggest we rename this test class to: `ConnectionManagementTest`

awoods commented:
Resolved with:
https://github.com/fcrepo4-exts/fcrepo-java-client/commit/d4197d12717e17cb41c807979b06719816655b26
https://github.com/fcrepo4-exts/fcrepo-java-client/commit/b8ccaaaad06b976e110a2df7d3b851c9f7c9ae93
https://github.com/fcrepo4-exts/fcrepo-java-client/commit/603179141af6e955771a34f60fc7a2bef28a2515
https://github.com/fcrepo4-exts/fcrepo-java-client/commit/de59b95e1de6924dd02bfb8e83762230dd8a66c5
